OpenRA.Mods.D2k/Traits/Buildings/ProductionFromMapEdge.cs

```diff
@@ -17,62 +18,100 @@
 namespace OpenRA.Mods.D2k.Traits
 {
 	[Desc("Produce a unit on the closest map edge cell and move into the world.")]
-	class ProductionFromMapEdgeInfo : ProductionInfo
+	class ProductionFromMapEdgeInfo : ProductionInfo, UsesInit<ProductionSpawnLocationInit>
 	{
 		public override object Create(ActorInitializer init) { return new ProductionFromMapEdge(init, this); }
 	}
 
 	class ProductionFromMapEdge : Production
 	{
+		readonly CPos spawnLocation;
+		readonly RallyPoint rp;
+		readonly DomainIndex domainIndex;
+
 		public ProductionFromMapEdge(ActorInitializer init, ProductionInfo info)
-			: base(init, info) { }
+			: base(init, info)
+		{
+			rp = init.Self.TraitOrDefault<RallyPoint>();
+			domainIndex = init.Self.World.WorldActor.Trait<DomainIndex>();
+
+			if (init.Contains<ProductionSpawnLocationInit>())
+				spawnLocation = init.Get<ProductionSpawnLocationInit, CPos>();
+		}
 
 		public override bool Produce(Actor self, ActorInfo producee, string factionVariant)
 		{
-			var location = self.World.Map.ChooseClosestEdgeCell(self.Location);
+			var aircraftInfo = producee.TraitInfoOrDefault<AircraftInfo>();
+			var mobileInfo = producee.TraitInfoOrDefault<MobileInfo>();
+
+			var passable = mobileInfo != null ? (uint)mobileInfo.GetMovementClass(self.World.TileSet) : 0;
+			var destination = rp != null ? rp.Location : self.Location;
+
+			var location = spawnLocation;
+			if (location == CPos.Zero)
```

Comment:
Relying on `CPos.Zero` as a magic value has caused lots of problems in the past.  Can we make spawnLocation a `CPos?` and check for `HasValue` instead?